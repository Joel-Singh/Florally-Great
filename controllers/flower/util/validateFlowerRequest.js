const asyncHandler = require("express-async-handler");
const path = require("path");
const Flower = require(path.join(appRoot, "models", "flower.js"));
const Region = require(path.join(appRoot, "models", "region.js"));
const { body } = require("express-validator");

const regionExistsValidation = asyncHandler(
  body("regionID").custom(async (regionID) => {
    const regionDocsWithIDs = await Region.find({}, "_id").exec();
    const regionIDArray = regionDocsWithIDs.map((doc) => doc._id.toString());

    if (!regionIDArray.includes(regionID))
      throw new Error("Region does not exist");
  })
);

const checkDuplicateFlower = asyncHandler(
  body("name").custom(async (name) => {
    const foundFlower = await Flower.findOne({ name }).exec();
    if (foundFlower !== null) {
      throw new Error("Flower with that name already exists");
    }
  })
);

module.exports = [
  body("name")
    .trim()
    .isLength({ min: 1 })
    .withMessage(`Name can't be empty`)
    .escape(),

  body("description")
    .trim()
    .isLength({ min: 1 })
    .withMessage(`Description can't be empty`)
    .escape(),

  body("numberInStock")
    .trim()
    .isLength({ min: 1 })
    .withMessage(`Number in stock can't be empty`)
    .isNumeric()
    .withMessage("Number in stock needs to be a number"),

  body("price")
    .trim()
    .isLength({ min: 1 })
    .withMessage(`Price can't be empty`)
    .matches(/\$[0-9]+\.[0-9][0-9]/)
    .withMessage("Price needs to be in $x.xx format, e.g $3.86 or $287.00"),

  regionExistsValidation,
  checkDuplicateFlower,
];
