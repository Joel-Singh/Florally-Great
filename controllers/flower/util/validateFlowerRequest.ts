import Flower from "../../../models/flower";
import Region from "../../../models/region";
import { body } from "express-validator";

const regionExistsValidation = body("regionID").custom(
  async (regionID: string) => {
    const regionDocsWithIDs = await Region.find({}, "_id").exec();
    const regionIDArray = regionDocsWithIDs.map((doc) => doc._id.toString());

    debugger;
    if (!regionIDArray.includes(regionID))
      throw new Error("Region does not exist");
  }
);

const checkDuplicateFlower = body("name").custom(async (name: string) => {
  const foundFlower = await Flower.findOne({ name }).exec();
  if (foundFlower !== null) {
    throw new Error("Flower with that name already exists");
  }
});

export default [
  body("name")
    .trim()
    .isLength({ min: 1 })
    .withMessage(`Name can't be empty`)
    .escape(),

  body("description")
    .trim()
    .isLength({ min: 1 })
    .withMessage(`Description can't be empty`)
    .escape(),

  body("numberInStock")
    .trim()
    .isLength({ min: 1 })
    .withMessage(`Number in stock can't be empty`)
    .isNumeric()
    .withMessage("Number in stock needs to be a number"),

  body("price")
    .trim()
    .isLength({ min: 1 })
    .withMessage(`Price can't be empty`)
    .matches(/\$[0-9]+\.[0-9][0-9]/)
    .withMessage("Price needs to be in $x.xx format, e.g $3.86 or $287.00"),

  regionExistsValidation,
  checkDuplicateFlower,
];
