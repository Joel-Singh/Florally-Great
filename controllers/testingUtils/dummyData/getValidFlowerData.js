const saveDummyRegion = require("./savingDummyDataToDb/saveDummyRegion.js");
const {
  generateSequentialNumber,
  generateSequentialObjectId,
} = require("./SequentialGenerators.js");

async function getValidFlowerPostData() {
  return {
    name: "Name" + generateSequentialNumber(),
    description: "Description",
    numberInStock: 32,
    price: "$3.89",
    regionID: (await saveDummyRegion())._id.toString(),
    _id: generateSequentialObjectId(),
  };
}

async function getValidFlowerModelData() {
  return {
    ...getValidFlowerPostData(),
    price: 3.89,
  };
}

module.exports = {
  getValidFlowerPostData,
  getValidFlowerModelData,
};
