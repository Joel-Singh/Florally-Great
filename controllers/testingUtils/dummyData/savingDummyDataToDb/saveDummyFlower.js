const path = require("path");
const Flower = require(path.join(appRoot, "models", "flower.js"));
const { generateSequentialObjectId } = require("../SequentialGenerators.js");
const saveDummyRegion = require("./saveDummyRegion.js");

module.exports = async function (overwrites = {}) {
  const dummyData = {
    name: "name",
    description: "description",
    price: 32,
    numberInStock: 21,
    region: (await saveDummyRegion())._id,
    _id: generateSequentialObjectId(),
    ...overwrites,
  };

  return await new Flower(dummyData).save();
};
