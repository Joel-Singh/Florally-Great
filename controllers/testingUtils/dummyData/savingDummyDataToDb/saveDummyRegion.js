const path = require("path");
const Region = require(path.join(appRoot, "models", "region.js"));
const { generateSequentialObjectId } = require("../SequentialGenerators.js");

module.exports = async function (overwrites = {}) {
  const dummyData = {
    name: "name",
    description: "description",
    _id: generateSequentialObjectId(),
    ...overwrites,
  };
  return await new Region(dummyData).save();
};
