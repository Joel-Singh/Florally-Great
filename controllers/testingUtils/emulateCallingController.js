module.exports = async function (controller, reqProperties, resProperties) {
  const { fakeReq, fakeRes } = getFakeMiddleware(reqProperties, resProperties);

  await controller(fakeReq, fakeRes);

  return {
    fakeReq,
    fakeRes,
    getRenderInformation,
  };

  function getFakeMiddleware(options) {
    const fakeReq = {
      ...reqProperties,
    };

    const fakeRes = {
      render: jest.fn(),
      redirect: jest.fn(),
      ...resProperties,
    };

    return { fakeReq, fakeRes };
  }
};

function getRenderInformation(fakeRes) {
  const renderCalls = fakeRes.render.mock.calls;
  if (renderCalls.length === 0) throw new Error("Render hasn't been called!");

  const renderCall = renderCalls[0];
  return {
    view: renderCall[0],
    locals: renderCall[1],
  };
}
