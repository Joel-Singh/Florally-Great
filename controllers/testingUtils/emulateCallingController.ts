import { Request, Response } from "express";

type nestedFunctionArray = Array<Function | nestedFunctionArray>;
export default async function emulateCallingController(
  controller: nestedFunctionArray | Function,
  reqProperties: Partial<Request> = {},
  resProperties: Partial<Response> = {},
) {
  const { fakeReq, fakeRes } = getFakeMiddleware(reqProperties, resProperties);
  const mockNext = jest.fn();

  if (!Array.isArray(controller)) {
    await controller(fakeReq, fakeRes, mockNext);
  } else {
    // @ts-ignore
    controller = controller.flat(Infinity) as Array<Function>;
    await runMiddlewareArray(
      controller as Array<Function>,
      fakeReq,
      fakeRes,
      mockNext,
    );
  }

  return {
    fakeReq,
    fakeRes,
    mockNext,
    getRenderInformation: createMockInfoGetter(
      fakeRes.render,
      "view",
      "locals",
    ),
    getRedirectInformation: createMockInfoGetter(
      fakeRes.redirect,
      "redirectPage",
    ),
    getMockNextInformation: createMockInfoGetter(mockNext, "errorOrEmpty"),
  };
}

function getFakeMiddleware(
  reqProperties: Partial<Request>,
  resProperties: Partial<Response>,
) {
  const fakeReq = {
    ...reqProperties,
  };

  const fakeRes = {
    ...resProperties,
    render: jest.fn(),
    redirect: jest.fn(),
    send: jest.fn(),
  };

  return { fakeReq, fakeRes };
}

function createMockInfoGetter<T extends Record<string, any>>(
  mockFunction: jest.Mock,
  ...namesInArgumentOrder: Array<keyof T>
) {
  return function () {
    const mockCalls = mockFunction.mock.calls;
    if (mockCalls.length === 0)
      throw new Error("Mock function hasn't been called!");

    const mockCall = mockCalls[0];
    const mockInfo: Partial<T> = {};

    namesInArgumentOrder.forEach((name, index) => {
      mockInfo[name] = mockCall[index];
    });

    return mockInfo as T;
  };
}

async function runMiddlewareArray(
  middlewares: Array<Function>,
  req: Partial<Request>,
  res: Partial<Response>,
  next: ReturnType<typeof jest.fn>,
) {
  for (const middleware of middlewares) {
    let previousCallLength = next.mock.calls.length;
    await middleware(req, res, next);
    let currentCallLength = next.mock.calls.length;
    let nextWasCalled = currentCallLength > previousCallLength;
    if (!nextWasCalled) break;
  }
}
