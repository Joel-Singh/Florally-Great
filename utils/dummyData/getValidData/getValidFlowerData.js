const {
  default: saveDummyRegion,
} = require("../savingDummyDataToDb/saveDummyRegion.ts");

const {
  generateSequentialNumber,
  generateSequentialObjectId,
} = require("../SequentialGenerators.ts");

async function getValidFlowerPostData() {
  return {
    name: "Name" + generateSequentialNumber(),
    description: "Description",
    numberInStock: 32,
    price: "$3.89",
    regionID: (await saveDummyRegion())._id.toString(),
    _id: generateSequentialObjectId(),
  };
}

async function getValidFlowerModelData() {
  const data = await getValidFlowerPostData();

  data.region = data.regionID;
  delete data.regionID;

  data.price = 3.89;

  return data;
}

module.exports = {
  getValidFlowerPostData,
  getValidFlowerModelData,
};
