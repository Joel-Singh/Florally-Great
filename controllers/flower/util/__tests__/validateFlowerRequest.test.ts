import { validationResult } from "express-validator";
import emulateCallingController from "../../../testingUtils/emulateCallingController";
import validateFlowerRequest from "../validateFlowerRequest";

import saveDummyRegion from "../../../../utils/dummyData/savingDummyDataToDb/saveDummyRegion";

import saveDummyFlower from "../../../../utils/dummyData/savingDummyDataToDb/saveDummyFlower";

import { getValidFlowerPostData } from "../../../../utils/dummyData/getValidData/getValidFlowerData";
import FlowerFormData from "../../../types/flowerFormData";

async function getValidationErrors(
  bodyOverwrites: Partial<FlowerFormData> = {},
  errorMsgMustInclude?: string
) {
  const { fakeReq } = await emulateCallingController(validateFlowerRequest, {
    body: {
      ...(await getValidFlowerPostData()),
      ...bodyOverwrites,
    },
  });

  const errors = validationResult(fakeReq).array();
  if (typeof errorMsgMustInclude === "undefined") return errors;
  else
    return errors.filter(({ msg }) => {
      return msg.includes(errorMsgMustInclude);
    });
}

async function testValidation<Key extends keyof FlowerFormData>(
  property: Key,
  propertyValue: FlowerFormData[Key],
  options:
    | {
        isValid: true;
      }
    | {
        isValid: false;
        errorFilter: string;
      }
) {
  if (options.isValid) {
    expect(
      await getValidationErrors({
        [property]: propertyValue,
      })
    ).toEqual([]);
  } else {
    expect(
      await getValidationErrors(
        {
          [property]: propertyValue,
        },
        options.errorFilter
      )
    ).toMatchSnapshot();
  }
}

test(`No errors with valid data`, async () => {
  expect(await getValidationErrors()).toEqual([]);
});

const flowerProperties: Array<keyof FlowerFormData> = [
  "name",
  "description",
  "numberInStock",
  "price",
];

describe("Empty input", () => {
  test.each(flowerProperties)(
    `Doesn't accept empty input for %s`,
    async (property) => {
      await testValidation(property, "", {
        isValid: false,
        errorFilter: "empty",
      });
    }
  );
});

describe("Just spaces", () => {
  test.each(flowerProperties)(
    `Doesn't accept just spaces for %s`,
    async (property) => {
      await testValidation(property, "   ", {
        isValid: false,
        errorFilter: "empty",
      });
    }
  );
});

test(`Only accepts numbers for number in stock`, async () => {
  await testValidation("numberInStock", "32", {
    isValid: true,
  });

  await testValidation("numberInStock", "not a number", {
    isValid: false,
    errorFilter: "needs to be a number",
  });
});

describe(`Price validation`, () => {
  const invalidFlowerPrices: string[] = ["$3aaa", "32.00", "32"];

  const validFlowerPrices: string[] = ["$3.86", "$3"];

  test.each(invalidFlowerPrices)(
    `%s flower price gives error`,
    async (flowerPrice) => {
      await testValidation("price", flowerPrice, {
        isValid: false,
        errorFilter: "format",
      });
    }
  );

  test.each(validFlowerPrices)(
    `%s flower price DOES NOT give error`,
    async (flowerPrice) => {
      await testValidation("price", flowerPrice, {
        isValid: true,
      });
    }
  );
});

test(`Doesn't accept nonexistent regions`, async () => {
  await testValidation("regionID", "asdfasdf", {
    isValid: false,
    errorFilter: "Region does not exist",
  });

  const realRegionID = (await saveDummyRegion())._id.toString();
  await testValidation("regionID", realRegionID, {
    isValid: true,
  });
});

test(`Doesn't accept duplicate flowers`, async () => {
  await saveDummyFlower({ name: "duplicate" });

  await testValidation("name", "duplicate", {
    isValid: false,
    errorFilter: "name already exists",
  });
});
