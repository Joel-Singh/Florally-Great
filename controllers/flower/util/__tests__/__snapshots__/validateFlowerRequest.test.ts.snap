// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Doesn't accept duplicate flowers 1`] = `
[
  {
    "location": "body",
    "msg": "Flower with that name already exists",
    "path": "name",
    "type": "field",
    "value": "duplicate",
  },
]
`;

exports[`Doesn't accept nonexistent regions 1`] = `
[
  {
    "location": "body",
    "msg": "Region does not exist",
    "path": "regionID",
    "type": "field",
    "value": "asdfasdf",
  },
]
`;

exports[`Empty input Doesn't accept empty input for description 1`] = `
[
  {
    "location": "body",
    "msg": "Description can't be empty",
    "path": "description",
    "type": "field",
    "value": "",
  },
]
`;

exports[`Empty input Doesn't accept empty input for name 1`] = `
[
  {
    "location": "body",
    "msg": "Name can't be empty",
    "path": "name",
    "type": "field",
    "value": "",
  },
]
`;

exports[`Empty input Doesn't accept empty input for numberInStock 1`] = `
[
  {
    "location": "body",
    "msg": "Number in stock can't be empty",
    "path": "numberInStock",
    "type": "field",
    "value": "",
  },
]
`;

exports[`Empty input Doesn't accept empty input for price 1`] = `
[
  {
    "location": "body",
    "msg": "Price can't be empty",
    "path": "price",
    "type": "field",
    "value": "",
  },
]
`;

exports[`Just spaces Doesn't accept just spaces for description 1`] = `
[
  {
    "location": "body",
    "msg": "Description can't be empty",
    "path": "description",
    "type": "field",
    "value": "",
  },
]
`;

exports[`Just spaces Doesn't accept just spaces for name 1`] = `
[
  {
    "location": "body",
    "msg": "Name can't be empty",
    "path": "name",
    "type": "field",
    "value": "",
  },
]
`;

exports[`Just spaces Doesn't accept just spaces for numberInStock 1`] = `
[
  {
    "location": "body",
    "msg": "Number in stock can't be empty",
    "path": "numberInStock",
    "type": "field",
    "value": "",
  },
]
`;

exports[`Just spaces Doesn't accept just spaces for price 1`] = `
[
  {
    "location": "body",
    "msg": "Price can't be empty",
    "path": "price",
    "type": "field",
    "value": "",
  },
]
`;

exports[`Only accepts numbers for number in stock 1`] = `
[
  {
    "location": "body",
    "msg": "Number in stock needs to be a number",
    "path": "numberInStock",
    "type": "field",
    "value": "not a number",
  },
]
`;

exports[`Price validation $3aaa flower price gives error 1`] = `
[
  {
    "location": "body",
    "msg": "Price needs to be in $x.xx format, e.g $3.86 or $287.00",
    "path": "price",
    "type": "field",
    "value": "$3aaa",
  },
]
`;

exports[`Price validation 32 flower price gives error 1`] = `
[
  {
    "location": "body",
    "msg": "Price needs to be in $x.xx format, e.g $3.86 or $287.00",
    "path": "price",
    "type": "field",
    "value": "32",
  },
]
`;

exports[`Price validation 32.00 flower price gives error 1`] = `
[
  {
    "location": "body",
    "msg": "Price needs to be in $x.xx format, e.g $3.86 or $287.00",
    "path": "price",
    "type": "field",
    "value": "32.00",
  },
]
`;
